### ISSUE TITLE
Support for nix/nixos in CrossJava

### ISSUE THREAD

**Issue by MasseGuillaume** [sbt/sbt#4260]
The current mechanism in CrossJava: https://github.com/sbt/sbt/blob/1.x/main/src/main/scala/sbt/internal/CrossJava.scala scan linux directories for java home.

As stated in the PR: https://github.com/sbt/sbt/pull/4245#issuecomment-404529319, there is no such default in nix/nixos. You need to provide the path explicitly.

Here is what I suggest:

Detect the user is using nix or nixos.

```scala
val isNixOS = sys.props.get("java.home").map(_.startsWith("/nix/store")).getOrElse(false)
```

Nixos users need to provide jdks via environment variables explicitly. Shortcircuit any automatic detection and use thos environment variable to build the map from version to jdk path.

```
SBT_JVM_XYZ=pkgs.jvm-xyz
```

**Comment by raboof** [comment]
It would be great to improve JDK detection on NixOS.

When using nix-shell it might be possible to auto-detect them from either the `PATH` or `buildInputs` environment variable?

**Comment by jorsn** [comment]
`JAVA_HOME` could help and should be more direct than parsing `PATH` or `buildInputs`.

It is defined in `pkgs/development/compilers/openjdk/8.nix` (likewise for oraclejdk):

```.sh
# Set JAVA_HOME automatically.
mkdir -p $out/nix-support
cat <<EOF > $out/nix-support/setup-hook
if [ -z "\$JAVA_HOME" ]; then export JAVA_HOME=$out/lib/openjdk; fi
EOF
```

Now, in a nix-shell:

```.sh
[user@host ~]$ nix-shell -p jdk
[nix-shell:~]$ echo $JAVA_HOME
/nix/store/5j2b3bnihr0rnd6gbn2999fbwxsj7b0k-openjdk-8u222-ga/lib/openjdk
[nix-shell:~]$ cat $JAVA_HOME/release 
JAVA_VERSION="1.8.0_222"
OS_NAME="Linux"
OS_VERSION="2.6"
OS_ARCH="amd64"
SOURCE=""

[nix-shell:~]$
```

**Comment by eed3si9n** [comment]
@jorsn The point of JDK forking is to detect and fork to JDK versions that are currently _not_ being used by JAVA_HOME though. Is there a way to enumerate all JDK packages and their metadata?

**Comment by jorsn** [comment]
Nix enables you to pull an arbitrary package as JDK into each build env, e.g. by simply spawning a nix-shell for each build (so, into `PATH`, setting `JAVA_HOME` automatically) and, if you like, to set `JAVA_HOME` to any JDK not in `PATH`.

But if you really need multiple JDKs visible *in one build*, then, in my opinion, there should not be a de-facto standard solely for the combination of Nix and sbt, but an easy way to give sbt an explicit list of `JAVA_HOME`s, e.g. as an env var `SBT_JAVA_HOMES`/`ALT_JAVA_HOMES` with colon separated directories (like `PATH`), a commandline option or a build property. sbt would only scan for available versions if they were not given.

This would at least generalize to any meta build tool managing dependencies (general package manager), which could safely control the collection of visible JDKs for sbt.

**Comment by raboof** [comment]
> there should not be a de-facto standard solely for the combination of Nix and sbt

Obviously!

> SBT_JAVA_HOMES

Let's not make this sbt-specific either ;)

> ALT_JAVA_HOMES with colon separated directories (like PATH)

Or perhaps simply `JAVA_HOMES`? I like this idea.

(it was one of the options I considered when creating https://github.com/NixOS/nixpkgs/pull/76699, but initially went with `JAVA_8_HOME` because there was more precedent for the latter, but both seem valid approaches to me)

> a commandline option or a build property

I'd prefer an environment variable.

**Comment by jorsn** [comment]
> > ALT_JAVA_HOMES with colon separated directories (like PATH)
> 
> Or perhaps simply `JAVA_HOMES`? I like this idea.

According to my intuition, `JAVA_HOMES` would include `JAVA_HOME` whereas `ALT_JAVA_HOMES` wouldn't.

**Comment by raboof** [comment]
> According to my intuition, JAVA_HOMES would include JAVA_HOME whereas ALT_JAVA_HOMES wouldn't.

Makes sense to me. I think I would prefer introducing `JAVA_HOMES` over `ALT_JAVA_HOMES`, that seems easier to keep consistent.
